paimon-core/src/main/java/org/apache/paimon/metrics/AbstractMetricGroup.java

```diff
@@ -0,0 +1,201 @@
+     * @param name name of the counter
+     * @return the created or existing counter
+     */
+    public Counter counter(String name) {
+        return counter(name, new SimpleCounter());
+    }
+
+    /**
+     * Registers a {@link org.apache.paimon.metrics.Counter}.
+     *
+     * @param name name of the counter
+     * @param counter counter to register
+     * @param <C> counter type
+     * @return the given counter
+     */
+    public <C extends Counter> C counter(String name, C counter) {
+        Metric metric = addMetric(name, counter);
+        return (C) metric;
+    }
+
+    /**
+     * Registers a new {@link org.apache.paimon.metrics.Gauge}.
+     *
+     * @param name name of the gauge
+     * @param gauge gauge to register
+     * @param <T> return type of the gauge
+     * @return the given gauge
+     */
+    public <T, G extends Gauge<T>> G gauge(String name, G gauge) {
+        Metric metric = addMetric(name, gauge);
+        return (G) metric;
+    }
+
+    /**
+     * Registers a new {@link Histogram} with Paimon.
+     *
+     * @param name name of the histogram
+     * @param histogram histogram to register
+     * @param <H> histogram type
+     * @return the registered histogram
+     */
+    public <H extends Histogram> H histogram(String name, H histogram) {
+        Metric metric = addMetric(name, histogram);
+        return (H) metric;
+    }
+
+    /**
+     * Adds the given metric to the group and registers it at the registry, if the group is not yet
+     * closed, and if no metric with the same name has been registered before.
+     *
+     * @param metricName the name to register the metric under
+     * @param metric the metric to register
+     */
+    protected Metric addMetric(String metricName, Metric metric) {
+        if (metric == null) {
+            LOG.warn(
+                    "Ignoring attempted registration of a metric due to being null for name {}.",
+                    metricName);
+            return null;
+        }
+        // add the metric only if the group is still open
+        if (!isClosed()) {
+            switch (metric.getMetricType()) {
+                case COUNTER:
+                case GAUGE:
+                case HISTOGRAM:
+                    // immediately put without a 'contains' check to optimize the common case
+                    // (no
+                    // collision)
+                    // collisions are resolved later
+                    Metric prior = metrics.put(metricName, metric);
+
+                    // check for collisions with other metric names
+                    if (prior != null) {
+                        // we had a collision. put back the original value
+                        metrics.put(metricName, prior);
+
+                        // we warn here, rather than failing, because metrics are tools that
+                        // should not
+                        // fail the
+                        // program when used incorrectly
```

Comment:
```suggestion
                        // we warn here, rather than failing, because metrics are tools that
                        // should not fail the program when used incorrectly
```

Reply:
Thanks for helping correct this.